crates/accelerate/src/synthesis/clifford/bm_synthesis.rs

```diff
@@ -0,0 +1,377 @@
+            clifford.tableau[[q1, q2 + 3]],
+            clifford.tableau[[q1 + 3, q2]],
+            clifford.tableau[[q1 + 3, q2 + 3]],
+        );
+        let mut mask = arr1(&[false, false, false, false, false, false]);
+        mask[q2] = true;
+        mask[q2 + 3] = true;
+        let xs = clifford.tableau.slice(s![q1, 0..6]).to_owned();
+        let zs = clifford.tableau.slice(s![q1 + 3, 0..6]).to_owned();
+        let ys = xs.clone() ^ zs.clone();
+        let is_loc_x = (xs.clone() & mask.clone()) == xs;
+        let is_loc_z = (zs.clone() & mask.clone()) == zs;
+        let is_loc_y = (ys.clone() & mask.clone()) == ys;
+        r1[[q1, q2]] = if is_loc_x || is_loc_z || is_loc_y {
+            1
+        } else {
+            0
+        } + if is_loc_x && is_loc_z && is_loc_y {
+            1
+        } else {
+            0
+        };
+    }
+
+    let mut diag1 = r1.diag().to_vec();
+    diag1.sort();
+    let mut diag2 = r2.diag().to_vec();
+    diag2.sort();
+
+    let nz1 = r1.iter().filter(|v| **v != 0).count();
+    let nz2 = r2.iter().filter(|v| **v != 0).count();
+
+    if diag1 == [2, 2, 2] {
+        0
+    } else if diag1 == [1, 1, 2] {
+        1
+    } else if (diag1 == [0, 1, 1])
+        || (diag1 == [1, 1, 1] && nz2 < 9)
+        || (diag1 == [0, 0, 2] && diag2 == [1, 1, 2])
+    {
+        2
+    } else if (diag1 == [1, 1, 1] && nz2 == 9)
+        || (diag1 == [0, 0, 1]
+            && (nz1 == 1 || diag2 == [2, 2, 2] || (diag2 == [1, 1, 2] && nz2 < 9)))
+        || (diag1 == [0, 0, 2] && diag2 == [0, 0, 2])
+        || (diag2 == [1, 2, 2] && nz1 == 0)
+    {
+        3
+    } else if diag2 == [0, 0, 1]
+        || (diag1 == [0, 0, 0]
+            && ((diag2 == [1, 1, 1] && nz2 == 9 && nz1 == 3)
+                || (diag2 == [0, 1, 1] && nz2 == 8 && nz1 == 2)))
+    {
+        5
+    } else if nz1 == 3 && nz2 == 3 {
+        6
+    } else {
+        4
+    }
+}
+
+/// Cost-reduction step. Given a Clifford over 2 or 3 qubits, finds a 2-qubit block
+/// (consisting of a single CX-gate, and two single-qubit I/V/W-gates applied to its
+/// control and target qubits) that reduces the two-qubit CX-cost of the Clifford
+/// (by definition, such a block must always exist).
+/// Returns the modified Clifford, the new cost (equal to the original cost - 1),
+/// and appends the corresponding gates to the ``gates`` vector. Slightly different
+/// from the Python implementation, the gates are already inverted (in practice this
+/// means applying an SGate instead of an SdgGate and vice versa), and at the end
+/// of the algorithm the vector of computed gates only needs to be reversed (but not
+/// also inverted).
+fn reduce_cost(
+    cliff: &Clifford,
+    cost: usize,
+    gates: &mut CliffordGatesVec,
+) -> Result<(Clifford, usize), String> {
+    // All choices for a 2-qubit block
+    for qubit0 in 0..cliff.num_qubits {
+        for qubit1 in qubit0 + 1..cliff.num_qubits {
+            for (n0, n1) in iproduct!(0..3, 0..3) {
```

Comment:
**Feedback wanted!** I very much like the `iproduct!` macro as it allows to write concisely nested `for` loops. Would it be possible to put `for` loops for `qubit0` and `qubit1` inside the same macro? Note that `qubit1` iterates from `qubit0+1`.

Reply:
With `iproduct` probably not as qubit0 and qubit1 are not independent... unless you'd want to do something like `iproduct!(0..n, 0..n, 0..3, 0..3).filter(|(q0, q1, n0, n1)| q1 > q0)` -- but it seems to me being explicit like you have it is benevolent for people reading the code in the future 😉